From 9bfbb39e65b76f4a20dca06610880cd1d661ef49 Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Thu, 4 Apr 2013 17:26:37 +0300
Subject: [PATCH 0539/1825] ALP: Switch information: Added Switch
 configuration, Updated board MAC info, and
 aligned Board GET functions

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 504427ff0302752b82b740966bc2ee0506474f26

- Added functions : mvBoard[ PhyAddrGet, SwitchCpuPortGet , SmiScanModeGet,
				SwitchIrqGet, SwitchPortMap , SwitchPortsMaskGet]

- Removed : mvBoardQuadPhyAddr0Get , mvBoardPhyLinkCryptPortAddrGet

Change-Id: I9e45fa048f4295438fe800d471dea698031d5b2e
Reviewed-on: http://vgitil04.il.marvell.com:8080/1434
Reviewed-by: Eran Ben-Avi <benavi@marvell.com>
Tested-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../avanta_lp_family/boardEnv/mvBoardEnvLib.c      |  145 ++++++++++++--------
 .../avanta_lp_family/boardEnv/mvBoardEnvLib.h      |   22 +--
 .../avanta_lp_family/boardEnv/mvBoardEnvSpec.c     |   30 +++-
 3 files changed, 121 insertions(+), 76 deletions(-)

diff --git a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
index 1329804..4952c94 100755
--- a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
+++ b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.c
@@ -302,13 +302,14 @@ MV_BOOL mvBoardIsPortInGmii(MV_U32 ethPortNum)
 *******************************************************************************/
 MV_32 mvBoardPhyAddrGet(MV_U32 ethPortNum)
 {
-	if (ethPortNum >= board->numBoardMacInfo)
-		return MV_ERROR;
 
 #if defined(CONFIG_MACH_AVANTA_LP_FPGA)
 	return 8;
 #endif
 
+	if (ethPortNum >= board->numBoardMacInfo)
+		return MV_ERROR;
+
 	return board->pBoardMacInfo[ethPortNum].boardEthSmiAddr;
 }
 
@@ -334,56 +335,6 @@ MV_BOARD_SPEC_INIT *mvBoardSpecInitGet(MV_VOID)
 }
 
 /*******************************************************************************
-* mvBoardQuadPhyAddr0Get - Get the phy address
-*
-* DESCRIPTION:
-*       This routine returns the Phy address of a given ethernet port.
-*
-* INPUT:
-*       ethPortNum - Ethernet port number.
-*
-* OUTPUT:
-*       None.
-*
-* RETURN:
-*       32bit describing Phy address, -1 if the port number is wrong.
-*
-*******************************************************************************/
-MV_32 mvBoardQuadPhyAddr0Get(MV_U32 ethPortNum)
-{
-#if !defined(CONFIG_MACH_AVANTA_LP_FPGA)
-	return board->pBoardMacInfo[ethPortNum].boardEthSmiAddr0;
-#else
-	return 0;
-#endif
-
-}
-
-/*******************************************************************************
-* mvBoardPhyLinkCryptPortAddrGet
-*
-* DESCRIPTION:
-*       This routine returns the Phy address of a given ethernet port.
-*
-* INPUT:
-*       ethPortNum - Ethernet port number.
-*
-* OUTPUT:
-*       None.
-*
-* RETURN:
-*       32bit describing Phy address, -1 if the port number is wrong.
-*
-*******************************************************************************/
-MV_32 mvBoardPhyLinkCryptPortAddrGet(MV_U32 ethPortNum)
-{
-	if (ethPortNum >= board->numBoardMacInfo)
-		return MV_ERROR;
-
-	return board->pBoardMacInfo[ethPortNum].LinkCryptPortAddr;
-}
-
-/*******************************************************************************
 * mvBoardMacSpeedGet - Get the Mac speed
 *
 * DESCRIPTION:
@@ -1106,6 +1057,80 @@ MV_STATUS mvBoardIsInternalSwitchConnected(MV_U32 ethPortNum)
 }
 
 /*******************************************************************************
+* mvBoardSwitchConnectedPortGet -
+*
+* DESCRIPTION:
+*       This routine returns the switch port connected to the ethPort
+*
+* INPUT:
+*       ethPortNum - Ethernet port number.
+*
+* OUTPUT:
+*       None.
+*
+* RETURN:
+*
+*******************************************************************************/
+MV_32 mvBoardSwitchConnectedPortGet(MV_U32 ethPort)
+{
+
+	if (board->switchInfoNum == 0)
+		return -1;
+
+	return board->pSwitchInfo[0].connectedPort[ethPort];
+}
+
+
+/*******************************************************************************
+* mvBoardSwitchPortsMaskGet -
+*
+* DESCRIPTION:
+*       This routine returns a mask describing all the connected switch ports
+*
+* INPUT:
+*       switchIdx - index of the switch. Only 0 is supported.
+*
+* OUTPUT:
+*       None.
+*
+* RETURN:
+*
+*******************************************************************************/
+MV_8 mvBoardSwitchPortsMaskGet(MV_U32 switchIdx)
+{
+	if ((board->switchInfoNum == 0) || (switchIdx >= board->switchInfoNum))
+	return -1;
+
+	return board->pSwitchInfo[switchIdx].connectedPortMask;
+}
+
+/*******************************************************************************
+* mvBoardInternalQuadPhyAddrGet - Get QUAD phy SMI address.
+*
+* DESCRIPTION:
+*       This routine returns the external QUAD phy address.
+*
+* INPUT:
+*       switchIdx - index of the switch. Only 0 is supported.
+*
+* OUTPUT:
+*       None.
+*
+* RETURN:
+*       The QUAD phy start address or -1 if error.
+*
+*******************************************************************************/
+MV_32 mvBoardInternalQuadPhyAddrGet(MV_U32 switchIdx)
+{
+
+	if ((board->switchInfoNum == 0) || (switchIdx >= board->switchInfoNum))
+		return -1;
+
+	return board->pSwitchInfo[switchIdx].internalQuadPhyAddr;
+
+}
+
+/*******************************************************************************
 * mvBoardSwitchPortForceLinkGet
 *
 * DESCRIPTION:
@@ -2244,7 +2269,10 @@ MV_STATUS mvBoardTwsiReadByteThruMux(MV_U8 muxChNum, MV_U8 chNum,
 *******************************************************************************/
 MV_32 mvBoardSmiScanModeGet(MV_U32 switchIdx)
 {
-	return board->pSwitchInfo[switchIdx].smiScanMode;
+	if ((board->switchInfoNum == 0) || (switchIdx >= board->switchInfoNum))
+		return -1;
+
+	return BOARD_ETH_SWITCH_SMI_SCAN_MODE;
 }
 
 /*******************************************************************************
@@ -2265,7 +2293,10 @@ MV_32 mvBoardSmiScanModeGet(MV_U32 switchIdx)
 *******************************************************************************/
 MV_32 mvBoardSwitchCpuPortGet(MV_U32 switchIdx)
 {
-	return 0;
+	if ((board->switchInfoNum == 0) || (switchIdx >= board->switchInfoNum))
+		return -1;
+
+	return board->pSwitchInfo[switchIdx].cpuPort;
 }
 
 /*******************************************************************************
@@ -2287,7 +2318,11 @@ MV_32 mvBoardSwitchCpuPortGet(MV_U32 switchIdx)
 *******************************************************************************/
 MV_32 mvBoardSwitchIrqGet(MV_VOID)
 {
-	return -1;
+
+	if (board->switchInfoNum == 0)
+		return -1;
+
+	return board->pSwitchInfo[0].switchIrq;
 }
 
 /*******************************************************************************
diff --git a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
index bba9521..8c5891b 100644
--- a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
+++ b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvLib.h
@@ -77,8 +77,8 @@ extern "C" {
 #include "boardEnv/mvBoardEnvSpec.h"
 #include "twsi/mvTwsi.h"
 
-#define BOARD_ETH_SWITCH_PORT_NUM       8
-#define BOARD_ETH_QD_SWITCH_PORT_NUM    5
+#define BOARD_ETH_SWITCH_PORT_NUM       7
+#define BOARD_ETH_SWITCH_SMI_SCAN_MODE	2
 #define MV_BOARD_MAX_MPP                9       /* number of MPP conf registers */
 #define MV_BOARD_MAX_MPP_GROUPS         9
 #define MV_BOARD_NAME_LEN               0x20
@@ -186,8 +186,8 @@ typedef struct _boardSwitchInfo {
 	MV_32 switchPort[BOARD_ETH_SWITCH_PORT_NUM];
 	MV_32 cpuPort;
 	MV_32 connectedPort[MV_ETH_MAX_PORTS];
-	MV_32 smiScanMode;
-	MV_32 quadPhyAddr[BOARD_ETH_QD_SWITCH_PORT_NUM];
+	MV_8 connectedPortMask;
+	MV_32 internalQuadPhyAddr;
 	MV_U32 forceLinkMask; /* Bitmask of switch ports to have force link (1Gbps) */
 } MV_BOARD_SWITCH_INFO;
 
@@ -242,9 +242,7 @@ typedef enum _boardMacSpeed {
 
 typedef struct _boardMacInfo {
 	MV_BOARD_MAC_SPEED boardMacSpeed;
-	MV_U8 boardEthSmiAddr;
-	MV_U16 LinkCryptPortAddr;
-	MV_U8 boardEthSmiAddr0;
+	MV_8 boardEthSmiAddr;
 } MV_BOARD_MAC_INFO;
 
 typedef struct _boardMppInfo {
@@ -401,7 +399,6 @@ typedef struct _boardInfo {
 #define MSAR_0_SPI1                             1
 
 /* For backward compatability with Legacy mode */
-#define mvBoardSwitchConnectedPortGet(port)     (-1)
 #define mvBoardIsSwitchConnected(port)          (mvBoardSwitchConnectedPortGet(port) != -1)
 /*#define mvBoardLinkStatusIrqGet(port)		mvBoardSwitchIrqGet()*/
 
@@ -416,9 +413,6 @@ MV_32 mvBoardSwitchPortMap(MV_U32 switchIdx, MV_U32 switchPortNum);
 MV_BOOL mvBoardIsPortLoopback(MV_U32 ethPortNum);
 MV_32 mvBoardPhyAddrGet(MV_U32 ethPortNum);
 MV_U8 mvBoardIoExpValGet(MV_BOARD_IO_EXPANDER_TYPE_INFO ioInfo);
-MV_32 mvBoardPhyLinkCryptPortAddrGet(MV_U32 ethPortNum);
-
-MV_32 mvBoardQuadPhyAddr0Get(MV_U32 ethPortNum);
 MV_STATUS mvBoardSarInfoGet(MV_SATR_TYPE_ID sarClass, MV_BOARD_SAR_INFO *sarInfo);
 MV_STATUS mvBoardConfigTypeGet(MV_CONFIG_TYPE_ID configClass, MV_BOARD_CONFIG_TYPE_INFO *configInfo);
 MV_STATUS mvBoardIoExpanderTypeGet(MV_IO_EXPANDER_TYPE_ID ioClass, MV_BOARD_IO_EXPANDER_TYPE_INFO *ioInfo);
@@ -469,12 +463,9 @@ MV_BOARD_BOOT_SRC mvBoardBootDeviceGet(MV_VOID);
 MV_U32 mvBoardBootAttrGet(MV_U32 sarBootDeviceValue, MV_U8 attrNum);
 MV_U8 mvBoardTwsiGet(MV_BOARD_TWSI_CLASS twsiClass, MV_U8 devNum, MV_U8 regNum);
 MV_STATUS mvBoardTwsiSet(MV_BOARD_TWSI_CLASS twsiClass, MV_U8 devNum, MV_U8 regNum, MV_U8 regVal);
-
 MV_U8 mvBoardCpuFreqGet(MV_VOID);
 MV_STATUS mvBoardCpuFreqSet(MV_U8 freqVal);
-
 MV_U8 mvBoardCpuCoresNumGet(MV_VOID);
-
 MV_STATUS mvBoardMppModulesScan(void);
 MV_STATUS mvBoardOtherModulesScan(void);
 MV_BOOL mvBoardIsPexModuleConnected(void);
@@ -495,6 +486,9 @@ MV_32 mvBoardGePhySwitchPortGet(void);
 MV_32 mvBoardRgmiiASwitchPortGet(void);
 MV_BOARD_MAC_SPEED mvBoardMacSpeedGet(MV_U32 ethPortNum);
 MV_32 mvBoardSwitchCpuPortGet(MV_U32 switchIdx);
+MV_32 mvBoardSwitchIrqGet(MV_VOID);
+MV_32 mvBoardSwitchConnectedPortGet(MV_U32 ethPort);
+MV_8 mvBoardSwitchPortsMaskGet(MV_U32 switchIdx);
 MV_BOOL mvBoardModuleAutoDetectEnabled(void);
 MV_32 mvBoardSmiScanModeGet(MV_U32 switchIdx);
 MV_BOARD_PEX_INFO *mvBoardPexInfoGet(void);
diff --git a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvSpec.c b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvSpec.c
index 5105989..9a49875 100755
--- a/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvSpec.c
+++ b/arch/arm/mach-avantalp/avanta_lp_family/boardEnv/mvBoardEnvSpec.c
@@ -226,9 +226,9 @@ MV_BOARD_TWSI_INFO db88f6660InfoBoardTwsiDev[] = {
 	{ BOARD_DEV_TWSI_IO_EXPANDER,	2,	0x24,	   ADDR7_BIT	},
 };
 MV_BOARD_MAC_INFO db88f6660InfoBoardMacInfo[] = {
-	/* {{MV_BOARD_MAC_SPEED boardMacSpeed, MV_U8 boardEthSmiAddr}} */
-	{ BOARD_MAC_SPEED_AUTO, 0x8									},
-	{ BOARD_MAC_SPEED_AUTO, 0x9									},
+	/* {{MV_BOARD_MAC_SPEED boardMacSpeed, MV_8 boardEthSmiAddr}} */
+	{ BOARD_MAC_SPEED_1000M, -1									},
+	{ BOARD_MAC_SPEED_AUTO, 0x1									},
 	{ N_A,			N_A									}
 };
 MV_BOARD_MPP_TYPE_INFO db88f6660InfoBoardModTypeInfo[] = {
@@ -253,6 +253,18 @@ MV_BOARD_MPP_GROUP_INFO db88f6660InfoBoardMppGroupConfig[] = {
 	  } }
 };
 
+MV_BOARD_SWITCH_INFO db88f6660InfoBoardSwitchValue[] = {
+	{
+	 .switchIrq = 29,	/* set to -1 for timer operation */
+	 .switchPort = {0, 1, 2, 3, 4, -1, -1},
+	 .cpuPort = 6,
+	 .connectedPort = {6, -1},
+	 .internalQuadPhyAddr = 0,
+	 .connectedPortMask= ( BIT0| BIT1| BIT2| BIT3| BIT4| BIT6),
+	 .forceLinkMask = 0x0
+	 }
+};
+
 MV_BOARD_INFO db88f6660_board_info = {
 	.boardName			= "DB-88F6660",
 	.numBoardMppTypeValue		= ARRSZ(db88f6660InfoBoardModTypeInfo),
@@ -293,6 +305,10 @@ MV_BOARD_INFO db88f6660_board_info = {
 	.gppPolarityValMid		= DB_88F6660_GPP_POL_MID,
 	.gppPolarityValHigh		= DB_88F6660_GPP_POL_HIGH,
 
+	/* External Switch Configuration */
+	.pSwitchInfo = db88f6660InfoBoardSwitchValue,
+	.switchInfoNum = ARRSZ(db88f6660InfoBoardSwitchValue),
+
 	/* TDM */
 	.numBoardTdmInfo		= {},
 	.pBoardTdmInt2CsInfo		= {},
@@ -325,10 +341,10 @@ MV_BOARD_INFO db88f6660_board_info = {
 
 MV_BOARD_MAC_INFO avanta_lp_customerInfoBoardMacInfo[] = {
 	/* {{MV_BOARD_MAC_SPEED	boardMacSpeed, MV_U8 boardEthSmiAddr}} */
-	{ BOARD_MAC_SPEED_AUTO, 0x10, 0x0 },
-	{ BOARD_MAC_SPEED_AUTO, 0x11, 0x0 },
-	{ BOARD_MAC_SPEED_AUTO, 0x12, 0x0 },
-	{ BOARD_MAC_SPEED_AUTO, 0x13, 0x0 }
+	{ BOARD_MAC_SPEED_AUTO, 0x10},
+	{ BOARD_MAC_SPEED_AUTO, 0x11},
+	{ BOARD_MAC_SPEED_AUTO, 0x12},
+	{ BOARD_MAC_SPEED_AUTO, 0x13}
 };
 
 MV_BOARD_MPP_TYPE_INFO avanta_lp_customerInfoBoardModTypeInfo[] = {
-- 
1.7.5.4

